### ISSUE TITLE
Duplicated entry in schema if casing of NamedSchema differ from filename

### ISSUE THREAD

**Issue by xavianaxw** [Redocly/redocly-cli#929]
**Describe the bug**

When creating a schema to validate an API request (Let's call it `ApiRequest`), if the casing for the NamedSchema is capitalize whereas the filename itself is not, it will result in a duplicate when running it through `redocly bundle`

**To Reproduce**

1. `.redocly.yaml`
```
extends: 
  - recommended

plugins: 
  -  src/schemata/plugins/extend-id-and-schema.js

apis:
  main@v0:
    root: src/schemata/openapi.yaml
    labels:
      - production
    rules:
      no-unused-components: warn

      # custom rule to ensure $id and $schema is passed
      # this is only enabled in external endpoints
      assert/required-id-and-schema:
        where:
          - subject: 
              type: Operation
              property: operationId
            assertions: 
              pattern: ^((?!(smoke|health)).)*$
          - subject:
              type: Response
              filterInParentKeys:
                - '200'
            assertions: 
              defined: true
        subject: 
          type: Schema
        assertions:
          defined: true
          required: 
            - $id
            - $schema
        message: Must have $id and $schema in Schema properties
        severity: error
```

2. `openapi.yaml`
```
openapi: 3.0.3
info:
  # info are hidden

servers:
  # servers are hidden

paths:
  # paths are hidden

components:
  schemas:
    ApiRequest:
      $ref: ./components/requests/apiRequest.yaml
```

and my `apiRequest.yaml`
```
$schema: http://json-schema.org/draft-07/schema
$id: ApiRequest
type: object
description: Request schema for /api/* endpoints

required:
  - jobId
  - hirerId
  - hirerOriginZone

properties:
  jobId:
    type: string
    description: Represent id of job ad
    example: 51563417

  hirerId:
    type: string
    description: Represent id of job ad's hirer
    example: 28829855

  hirerOriginZone:
    type: string
    description: Represent origin zone of hirer's location.
    example: anz-1
```

3. Run this command
```
redocly bundle ./src/schemata/openapi.yaml -o ./src/schemata/openapi-generated --ext=json --format=json  --dereferenced --lint --skip-rule=spec
```

4. Output
```
{
  "openapi": "3.0.3",
  # stripped out info
  # stripped out paths
  "components": {
    "schemas": {
      "ApiRequest": {
        "$schema": "http://json-schema.org/draft-07/schema",
        "$id": "ApiRequest",
        "type": "object",
        "description": "Request schema for /api/* endpoints",
        "required": [
          "jobId",
          "hirerId",
          "hirerOriginZone"
        ],
        "properties": {
          "jobId": {
            "type": "string",
            "description": "Represent id of job ad",
            "example": 51563417
          },
          "hirerId": {
            "type": "string",
            "description": "Represent id of job ad's hirer",
            "example": 28829855
          },
          "hirerOriginZone": {
            "type": "string",
            "description": "Represent origin zone of hirer's location.",
            "example": "anz-1"
          }
        }
      },
      "apiRequest": {
        "$schema": "http://json-schema.org/draft-07/schema",
        "$id": "ApiRequest",
        "type": "object",
        "description": "Request schema for /api/* endpoints",
        "required": [
          "jobId",
          "hirerId",
          "hirerOriginZone"
        ],
        "properties": {
          "jobId": {
            "type": "string",
            "description": "Represent id of job ad",
            "example": 51563417
          },
          "hirerId": {
            "type": "string",
            "description": "Represent id of job ad's hirer",
            "example": 28829855
          },
          "hirerOriginZone": {
            "type": "string",
            "description": "Represent origin zone of hirer's location.",
            "example": "anz-1"
          }
        }
      }
    }
  }
}
```

**Expected behavior**
```
{
  "openapi": "3.0.3",
  # strip out info
  # strip out paths
  "components": {
    "schemas": {
      "ApiRequest": {
        "$schema": "http://json-schema.org/draft-07/schema",
        "$id": "ApiRequest",
        "type": "object",
        "description": "Request schema for /api/* endpoints",
        "required": [
          "jobId",
          "hirerId",
          "hirerOriginZone"
        ],
        "properties": {
          "jobId": {
            "type": "string",
            "description": "Represent id of job ad",
            "example": 51563417
          },
          "hirerId": {
            "type": "string",
            "description": "Represent id of job ad's hirer",
            "example": 28829855
          },
          "hirerOriginZone": {
            "type": "string",
            "description": "Represent origin zone of hirer's location.",
            "example": "anz-1"
          }
        }
      }
    }
  }
}
```

**Logs**
![Screen Shot 2022-11-10 at 4 02 02 pm](https://user-images.githubusercontent.com/8082158/201004984-736c2e98-02f5-4229-b24f-0e6f73d84ce1.png)

**OpenAPI definition**
OAS 3.0

**Redocly Version(s)**
`1.0.0-beta.112`

**`Node.js` Version(s)**
`v16.17.0`

**Additional context**
tldr; Expecting one output of `ApiRequest` (following casing of NamedSchema) instead of having 2 where there other is a lowercase (when the filename starts with a lowercase)

> Weird behaviour but when I do `$ref: '../components/requests/ApiRequest.yaml` it does not throw an error in `lint` nor the vscode extension

**Comment by tatomyr** [comment]
Hi @xavianaxw,

Thanks for reporting this.

It is because at first we resolve the file into `components` and then replace the external reference with the internal one. You've run `bundle` command with the `dereference` option so the internal ref also got resolved but the unused file-named-component remained. 

You may try using `--remove-unused-components` option to get rid of the duplicates. Or alternatively, you might use a [custom decorator](https://redocly.com/docs/cli/decorators/#custom-decorators) to remove the exact reference you want.

**Comment by xavianaxw** [comment]
Let me give it a go. Will report back

**Comment by xavianaxw** [comment]
Did some playtesting and here's what I find.

- Using `redocly lint --dereferenced --remove-unused-components` will remove all `schemas` as they are unused.
- Using `redocly lint --remove-unused-components` will keep all the schemas I used in `paths` except the `ApiRequest` that was declared exclusively in `openapi.yaml`

Both doesn't do what I require.

**Comment by adamaltman** [comment]
Did you try without `--dereferenced`?

**Comment by xavianaxw** [comment]
@adamaltman yeah I did `redocly lint --remove-unused-components`

It keeps all the schemas I've used in `paths` but since `ApiRequest` was declared in `openapi.yaml` like so:

```
components:
  schemas:
    ApiRequest:
      $ref: ./components/requests/apiRequest.yaml
```

It is treated as unused and thus removed from my final build

**Comment by kanoru3101** [comment]
A couple of things have changed since this was opened. 
`bundle` now has a `--component-names-strategy=title` option, so components are named after the schema's `title` rather than the file. In your case that means adding `title: ApiRequest` to `apiRequest.yaml` and running `redocly bundle openapi.yaml -o openapi.json --component-names-strategy=title`. See the [docs](https://redocly.com/docs/cli/commands/bundle#configure-the-component-names-strategy).

The simpler fix is to just rename the file to `ApiRequest.yaml`, then the names match and there's no duplicate.
